New package `alfa-painting-order`

rcj-siteimprove opened the pull request:
This PR introduces a new package which contains functionality for computing the painting order of HTML elements based on the specification: https://drafts.csswg.org/css2/#painting-order.

The process for computing the painting order is fairly complex and that reflects in the implementation. This is the first step to make something that works in most common use cases. There are possibly room for caching some computations and other optimizations as well as making the algorithm easier to follow and less error-prone.

The current implementation does not consider scrolling frames, but that could be a logical next step if needed.

The painting order can be used for e.g. detecting if elements are fully or partially covered by other elements, which is relevant when determining the clickable area of an element. More on that in future PRs.

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,241 @@
+      }
+    }
+
+    for (const blockLevel of blockLevels) {
+      if (!defer && createsSC(blockLevel)) {
+        paint(blockLevel, canvas);
+      } else {
+        canvas.push(blockLevel);
+      }
+    }
```
**Suggestion:** It looks like this loop could be refactored by abstracting the array it traverses and the second part of the predicate.
The `positionedOrStackingContexts` can also be split at negative `z-index` and thus the first and last loops also fit in that refactoring.

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,241 @@
+    },
+  ): void {
+    const { defer = false } = options;
+    const positionedOrStackingContexts: Array<Element> = [];
+    const blockLevels: Array<Element> = [];
+    const floats: Array<Element> = [];
+    const inlines: Array<Element> = [];
```
**Suggestion:** It feels like we could use an abstraction ("layer"?) to wrap these. Then `distributeIntoLayers` would produce that abstraction and the next step of pushing things in the `canvas` is a `.toPaintingOrder` (or `.fromLayer`) builder.

rcj-siteimprove replied:
I tried to get the abstraction right and I think it should be doable, but I'd like to do it however as part of a future improvement PR. I think, in any case, that I am going to improve on this once we start using it.

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,241 @@
+        .filter(and(Element.isElement, rendered))) {
+        distributeIntoLayers(child);
+
+        if (or(isPositioned, isFloat, createsSC)(child)) {
```
**Suggestion:** It seems that this test is more or less the same that the one(s) in `distributeIntoLayers` that decide to recurse in the subtree. It feels smelly to have that in two places (and error prone if we change one), especially given that they are not written exactly the same way. Not sure how we can improve that, though.

rcj-siteimprove replied:
Same as the above comment, I'd like to improve on this and make the code better structured in the near future.

Jym77 reviewed (changes requested):
Looks good. Some code organisation comments.

rcj-siteimprove commented:
!pr extract

rcj-siteimprove commented:
!pr extract

rcj-siteimprove commented:
@Jym77 I addressed most of the comments, except two that I would like to look into in a follow up PR, but after the other work that depend on this one (using it to compute clickable area).

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,286 @@
+  protected constructor(elements: Array<Element>) {
+    this._elements = elements;
+  }
+
```
```suggestion

public get elements(): Iterable<Element> {
  return this._elements;
}

```

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,286 @@
+      (PaintingOrder.isPaintingOrder(value) &&
+        Array.equals(value._elements, this._elements))
+    );
+  }
+  public hash(hash: Hash): void {
```
```suggestion
  }
  
  public hash(hash: Hash): void {
```

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,286 @@
+  public equals(value: unknown): value is this;
+  public equals(value: unknown): boolean {
+    return (
+      value === this ||
```
```suggestion
```
We usually don't have the physical equality test in our `.equals` (which does indeed make them less efficient in that case, but that is hopefully a rare case 🤔

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,286 @@
+    return PaintingOrder.of(paint(device, root, Sequence.empty()));
+  });
+
+  const getZLevel = Cache.memoize(function (device: Device, element: Element) {
```
**Suggestion:** This looks cheap to compute (because computed styles are already cached) and probably doesn't need to be cached.

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,286 @@
+    device: Device,
+    element: Element,
+    canvas: Sequence<Element>,
+    options: { defer?: boolean } = {
+      defer: false,
+    },
+  ): Sequence<Element> {
+    const { defer = false } = options;
```
**Suggestion:** This defines the default for `defer` twice, which can get confusing. E.g. if the second one changes to `const { defer = true } = options`, it requires `options` to be `{}` in order to trigger (`undefined` will get the other default value).

Jym77 commented on packages/alfa-painting-order/src/painting-order.ts:
```diff
@@ -0,0 +1,286 @@
+      positionedOrStackingContexts,
+    );
+
+    canvas = paintLayer(device, canvas, negatives, element, defer);
+    canvas = paintLayer(device, canvas, blockLevels, element, defer);
+    canvas = paintLayer(device, canvas, floats, element, defer);
+    canvas = paintLayer(device, canvas, inlines, element, defer);
+    canvas = paintLayer(device, canvas, nonNegatives, element, defer);
```
**Note:** It looks like `paintLayer` is only adding to the canvas, not changing more. So it could just return the new elements and this would be a big `canvas.concat`.

Probably for another time, though.

rcj-siteimprove replied:
That makes sense, I'll keep that in mind for the future.

rcj-siteimprove commented:
!pr extract
